source/common/router/router.cc

```diff
@@ -423,20 +441,25 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
     do_shadowing_ = false;
   }
 
-  if (buffering) {
-    // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
-    // since it's all moves from here on.
-    Buffer::OwnedImpl copy(data);
-    upstream_request_->encodeData(copy, end_stream);
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    UpstreamRequest* upstream_request = it->get();
+    if (buffering) {
+      // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
```

Comment:
Do we need to make a copy in the non-buffering case too now that we're passing `data` to multiple `encodeData`?

Reply:
yeah makes sense, should I avoid copying if there is 1 upstream request and not buffering?